c error when wrapping a function in an optional function type

### Describe the bug

`?Foo(none)` with `Foo` being a function type def causes a build error

### Reproduction Steps

`v -cg -show-c-output -cc gcc .`

```v
type Foo = fn (bar string)

fn main() {
    mut func := ?Foo(none)
    dump(func)
}
```

### Expected Behavior

a compiled program

### Current Behavior

```
======== Output of the C Compiler (gcc) ========
C:\Users\nicof\AppData\Local\Temp\v_0\foo.01JB7DV7N82CMDR7ZC0RYB36A4.tmp.c: In function '_v_dump_expr__option_main__Foo':
C:\Users\nicof\AppData\Local\Temp\v_0\foo.01JB7DV7N82CMDR7ZC0RYB36A4.tmp.c:2305:24: error: too few arguments to function '_option_anon_fn_string_str'
 2305 |         string value = _option_anon_fn_string_str();
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\nicof\AppData\Local\Temp\v_0\foo.01JB7DV7N82CMDR7ZC0RYB36A4.tmp.c:2291:8: note: declared here
 2291 | string _option_anon_fn_string_str(_option_anon_fn_string it) { return indent__option_anon_fn_string_str(it, 0); }
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\nicof\AppData\Local\Temp\v_0\foo.01JB7DV7N82CMDR7ZC0RYB36A4.tmp.c: In function 'main__main':
C:\Users\nicof\AppData\Local\Temp\v_0\foo.01JB7DV7N82CMDR7ZC0RYB36A4.tmp.c:13645:75: error: incompatible type for argument 4 of '_v_dump_expr__option_main__Foo'
13645 |          _v_dump_expr__option_main__Foo(_SLIT("foo.v"), 5, _SLIT("func"), func);
      |                                                                           ^~~~
      |                                                                           |
      |                                                                           _option_anon_fn_string
C:\Users\nicof\AppData\Local\Temp\v_0\foo.01JB7DV7N82CMDR7ZC0RYB36A4.tmp.c:2303:128: note: expected 'DumpFNType__option_main__Foo' {aka 'void (*)(string)'} but argument is of type '_option_anon_fn_string'
 2303 | DumpFNType__option_main__Foo _v_dump_expr__option_main__Foo(string fpath, int line, string sexpr, DumpFNType__option_main__Foo dump_arg) {
      |                                                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
================================================
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.8 c8ddf21

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.8 c8ddf21
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: C:\Users\nicof\Desktop\Projects\V\studytui
vexe: C:\Users\nicof\Desktop\v\v.exe
vexe mtime: 2024-10-27 16:43:14

vroot: OK, value: C:\Users\nicof\Desktop\v
VMODULES: OK, value: C:\Users\nicof\.vmodules
VTMP: OK, value: C:\Users\nicof\AppData\Local\Temp\v_0

Git version: git version 2.45.2.windows.1
Git vroot status: weekly.2024.43-47-gc8ddf219
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc: N/A
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzFlNmY0YjI3Y2M3NjgxYzYyMTBlMTUiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.d1-vighmDWrSsVjeKeCBMtHCbqweDbun2ONcPMvGUuI">Huly&reg;: <b>V_0.6-21119</b></a></sub>

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -106,20 +106,26 @@ fn (mut g Gen) dump_expr_definitions() {
 		is_ptr := typ.is_ptr()
 		deref, _ := deref_kind(str_method_expects_ptr, is_ptr, typ)
 		to_string_fn_name := g.get_str_fn(typ.clear_flags(.shared_f, .result))
+		is_option := typ.has_option_or_result()
 		mut ptr_asterisk := if is_ptr { '*'.repeat(typ.nr_muls()) } else { '' }
 		mut str_dumparg_type := ''
 		mut str_dumparg_ret_type := ''
 		if dump_sym.kind == .none {
 			str_dumparg_type = 'IError' + ptr_asterisk
+		} else if dump_sym.kind == .function {
+			if is_option {
+				ptr_asterisk = ptr_asterisk.replace('*', '_ptr')
+			}
+			str_dumparg_type += g.styp(typ).replace('*', '') + ptr_asterisk
 		} else {
-			if typ.has_flag(.option) {
+			if is_option {
 				str_dumparg_type += '_option_'
 				ptr_asterisk = ptr_asterisk.replace('*', '_ptr')
 			}
 			str_dumparg_type += g.cc_type(typ, true) + ptr_asterisk
 		}
 		mut is_fixed_arr_ret := false
-		if dump_sym.kind == .function {
+		if dump_sym.kind == .function && !is_option {
 			fninfo := dump_sym.info as ast.FnType
 			str_dumparg_type = 'DumpFNType_${name}'
 			tdef_pos := g.out.len
@@ -128,7 +134,7 @@ fn (mut g Gen) dump_expr_definitions() {
 			g.go_back(str_tdef.len)
 			dump_typedefs['typedef ${str_tdef};'] = true
 			str_dumparg_ret_type = str_dumparg_type
-		} else if !typ.has_option_or_result() && dump_sym.is_array_fixed() {
+		} else if !is_option && dump_sym.is_array_fixed() {
 			match dump_sym.kind {
 				.array_fixed {
 					if (dump_sym.info as ast.ArrayFixed).is_fn_ret {
@@ -172,7 +178,7 @@ fn (mut g Gen) dump_expr_definitions() {
 		}
 		mut surrounder := util.new_surrounder(3)
 		surrounder.add('\tstring sline = int_str(line);', '\tstring_free(&sline);')
-		if dump_sym.kind == .function {
+		if dump_sym.kind == .function && !is_option {
 			surrounder.add('\tstring value = ${to_string_fn_name}();', '\tstring_free(&value);')
 		} else if dump_sym.kind == .none {
 			surrounder.add('\tstring value = _SLIT("none");', '\tstring_free(&value);')

--- a/vlib/v/tests/options/option_fn_alias_test.v
+++ b/vlib/v/tests/options/option_fn_alias_test.v
@@ -1,0 +1,12 @@
+type Foo = fn (bar string)
+
+fn test_main() {
+	mut func := ?Foo(none)
+	res := dump(func)
+	assert res == none
+
+	mut func2 := ?Foo(fn (bar string) {})
+	res2 := dump(func2)
+	assert res2 != none
+	assert res2?.str() == 'fn (string)'
+}

